src/sctools/groups.py

```diff
@@ -0,0 +1,151 @@
+"""
+Group QC outputs
+
+"""
+
+from crimson import picard
+import argparse
+import os
+import pandas as pd
+
+
+def aggregate_picard_metrics_by_row(file_names, output_name):
+    """
+    pipeline output picard QC metrics at single cell/sample level.
+    This function is called to merge/aggregate QC metrics by metrics type and
+    then merge multiple QC measurement into single matrix file.
+    In this file, column is sample/cell and row is QC metrics
+    :param file_names: list of files generated by Picard
+    :param output_name: output name without suffix'
+    """
+    # initial output
+    metrics = {}
+    d = pd.DataFrame()
+    for file_name in file_names:
+        cell_id = os.path.basename(file_name).split('_qc')[0]
+        metrics[cell_id] = {}
+        parsed = picard.parse(file_name)
+        class_name = parsed['metrics']['class'].split('.')[2]
+        # Aignment metrics return multiple lines,
+        # but only output PAIRED-READS/third line
+        contents = parsed['metrics']['contents']
+        if class_name == "AlignmentSummaryMetrics":
+            # parse out PE, R1 and R2
+            rows = {}
+            for m in contents:
+                cat = m['CATEGORY']
+                rows.update({k + '.' + cat: v for k, v in m.items() if k not in ['SAMPLE', 'LIBRARY', 'READ_GROUP', 'CATEGORY']})
+        # sometimes(very rare), insertion metrics also return multiple lines
+        # results to include TANDEM repeats. but we only output the first line.
+        elif class_name == "InsertSizeMetrics":
+            # if the elemnet counts is less than 21,
+            # it means insertion metrics returns multiple line results.
+            if len(contents) < 21:
+                rows = contents[0]
+            else:
+                rows = contents
+        else:
+            # other metrics(so far) only return one line results.
+            rows = contents
+        metrics[cell_id].update({
+                k: rows[k] for k in rows if k not in
+                ['SAMPLE', 'LIBRARY', 'READ_GROUP', 'CATEGORY']
+                })
+        df = pd.DataFrame.from_dict(metrics, orient='columns')
+        df.insert(0, 'Class', class_name)
+        d = d.append(df)
+    d.to_csv(output_name + '.csv')
+
+
+def aggregate_picard_metrics_by_table(file_names, output_name):
+    """
+    pipeline output picard QC metrics at single cell/sample level.
+    This function is called to merge/aggregate QC metrics by metrics type and
+    then merge multiple QC measurement into single matrix file.
+    In this file, column is sample/cell and row is QC metrics
```

Comment:
Would prefer this be transposed. What was the motivation for this orientation?

Reply:
the output of this way would be "a long skinny" table(not vector). The first 2 or 3 columns are meta data related to Picard metrics and the last column is the actual measurement. I think it is more traditional way to output Picard metrics.

Reply:
Ok, my motivation is that consistency of our object model is important. In our matrix discussions, we came to the conclusion that features/variables as columns and observations as rows was what we'd like to go with. 

Transposing this matrix would be consistent with that. I don't find the long vs. tall discussion compelling because a small SS2 experiment also outputs a long matrix (~1000 cells x ~40,000-60,000 features).

Reply:
we extract two different metrics from Picard outputs, one is the row-based and another is the table-based. For the row based, it is certainly we can transpose them into any ways. For the table-based, such as base distribution, gc bias distribution, error distribution.... it probably is not necessary to transpose them at all.

Reply:
changed